consensus/src/dag/adapter.rs

```diff
@@ -87,3 +98,164 @@ impl Notifier for BufferManagerAdapter {
+                new_block_event.round(),
+                new_block_event.proposer(),
+            ),
+            Self::bitvec_to_validators(
+                validators,
+                &new_block_event.previous_block_votes_bitvec().clone().into(),
+            )?,
+            Self::indices_to_validators(validators, new_block_event.failed_proposer_indices())?,
+        ))
+    }
+}
+
+impl DAGStorage for StorageAdapter {
+    fn save_pending_node(&self, node: &Node) -> anyhow::Result<()> {
+        Ok(self.consensus_db.put::<NodeSchema>(&(), node)?)
+    }
+
+    fn get_pending_node(&self) -> anyhow::Result<Option<Node>> {
+        Ok(self.consensus_db.get::<NodeSchema>(&())?)
+    }
+
+    fn delete_pending_node(&self) -> anyhow::Result<()> {
+        Ok(self.consensus_db.delete::<NodeSchema>(vec![()])?)
+    }
+
+    fn save_vote(&self, node_id: &NodeId, vote: &Vote) -> anyhow::Result<()> {
+        Ok(self.consensus_db.put::<DagVoteSchema>(node_id, vote)?)
+    }
+
+    fn get_votes(&self) -> anyhow::Result<Vec<(NodeId, Vote)>> {
+        Ok(self.consensus_db.get_all::<DagVoteSchema>()?)
+    }
+
+    fn delete_votes(&self, node_ids: Vec<NodeId>) -> anyhow::Result<()> {
+        Ok(self.consensus_db.delete::<DagVoteSchema>(node_ids)?)
+    }
+
+    fn save_certified_node(&self, node: &CertifiedNode) -> anyhow::Result<()> {
+        Ok(self
+            .consensus_db
+            .put::<CertifiedNodeSchema>(&node.digest(), node)?)
+    }
+
+    fn get_certified_nodes(&self) -> anyhow::Result<Vec<(HashValue, CertifiedNode)>> {
+        Ok(self.consensus_db.get_all::<CertifiedNodeSchema>()?)
+    }
+
+    fn delete_certified_nodes(&self, digests: Vec<HashValue>) -> anyhow::Result<()> {
+        Ok(self.consensus_db.delete::<CertifiedNodeSchema>(digests)?)
+    }
+
+    fn save_ordered_anchor_id(&self, node_id: &NodeId) -> anyhow::Result<()> {
+        Ok(self
+            .consensus_db
+            .put::<OrderedAnchorIdSchema>(node_id, &())?)
+    }
+
+    fn get_ordered_anchor_ids(&self) -> anyhow::Result<Vec<(NodeId, ())>> {
+        Ok(self.consensus_db.get_all::<OrderedAnchorIdSchema>()?)
+    }
+
+    fn delete_ordered_anchor_ids(&self, node_ids: Vec<NodeId>) -> anyhow::Result<()> {
+        Ok(self
+            .consensus_db
+            .delete::<OrderedAnchorIdSchema>(node_ids)?)
+    }
+
+    fn get_latest_k_committed_events(&self, k: u64) -> anyhow::Result<Vec<CommitEvent>> {
+        let latest_db_version = self.aptos_db.get_latest_version().unwrap_or(0);
+        let mut commit_events = vec![];
+        for event in self.aptos_db.get_events(
+            &new_block_event_key(),
+            u64::MAX,
+            Order::Descending,
+            k as u64,
+            latest_db_version,
+        )? {
+            let new_block_event = bcs::from_bytes::<NewBlockEvent>(event.event.event_data())?;
+            if self
+                .epoch_to_validators
+                .contains_key(&new_block_event.epoch())
```

Comment:
What if we do not have this epoch? What does it mean?

Reply:
it's guaranteed to have the current epoch, the map is for past epochs

Reply:
but, this should be an assertion right? otherwise we can have different vec of commit_events.